can/notifier.py

```diff
@@ -24,37 +36,40 @@ def __init__(self, bus, listeners, timeout=1.0, loop=None):
             many listeners carry out flush operations to persist data.
 
 
-        :param can.BusABC bus: A :ref:`bus` or a list of buses to listen to.
-        :param list listeners: An iterable of :class:`~can.Listener`
-        :param float timeout: An optional maximum number of seconds to wait for any message.
-        :param asyncio.AbstractEventLoop loop:
-            An :mod:`asyncio` event loop to schedule listeners in.
+        :param bus: A :ref:`bus` or a list of buses to listen to.
+        :param listeners: An iterable of :class:`~can.Listener`
+        :param timeout: An optional maximum number of seconds to wait for any message.
+        :param loop: An :mod:`asyncio` event loop to schedule listeners in.
         """
-        self.listeners = listeners
+        self.listeners = list(listeners)
         self.bus = bus
         self.timeout = timeout
         self._loop = loop
 
         #: Exception raised in thread
-        self.exception = None
+        self.exception: Optional[Exception] = None
 
         self._running = True
         self._lock = threading.Lock()
 
-        self._readers = []
+        self._readers: List[Union[int, threading.Thread]] = []
         buses = self.bus if isinstance(self.bus, list) else [self.bus]
         for bus in buses:
             self.add_bus(bus)
 
-    def add_bus(self, bus):
+    def add_bus(self, bus: BusABC):
         """Add a bus for notification.
 
-        :param can.BusABC bus:
+        :param bus:
             CAN bus instance.
         """
-        if self._loop is not None and hasattr(bus, "fileno") and bus.fileno() >= 0:
+        if (
+            self._loop is not None
+            and hasattr(bus, "fileno")
+            and bus.fileno() >= 0  # type: ignore
```

Comment:
This is disabled here because mypy doesn't really deal well with `hasattr` checks. See the discussion [here](https://github.com/python/mypy/issues/1424). We could either:

1. Disable the type checker (like is done here)
2. Perform the cast to `Any`
3. Refactor the code to have `fileno` on the base class return `None` by default (which we can check for), and then subclasses provide their own implementation as needed

I'm personally inclined to 3, but that's a bit more involved than the other 2 (and perhaps should be kept out of this change). Thoughts?

Reply:
I vote for 3, but it can return -1 instead of None.

Reply:
I'd vote for 1. It's just a shortcoming of the type system, and no problem with our API.

Reply:
I agree that one makes sense for this PR, but 3 would be the best long term solution.